File: library/src/main/java/org/incendo/interfaces/grid/GridPoint.java
package org.incendo.interfaces.grid;

public record GridPoint(int x, int y) {

    public static GridPoint at(final int x, final int y) {
        return new GridPoint(x, y);
    }

    public GridPoint minus(final GridPoint other) {
        return new GridPoint(this.x - other.x(), this.y - other.y());
    }
}

File: next/src/main/java/org/incendo/interfaces/next/grid/GridPoint.kt
package org.incendo.interfaces.next.grid

/**
 * A 2-dimensional vector storing integer components.
 */
public data class GridPoint(val x: Int, val y: Int) {

    public companion object {
        public fun at(x: Int, y: Int): GridPoint = GridPoint(x, y)
    }

    public operator fun minus(other: GridPoint): GridPoint {
        return GridPoint(x - other.x, y - other.y)
    }

    public operator fun rangeTo(other: GridPoint): Sequence<GridPoint> = sequence {
        for (y in y..other.y) {
            for (x in x..other.x) {
                yield(at(x, y))
            }
        }
    }
}
